torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -138,7 +138,116 @@ def get_tp_parallel_strategy(
     return RowwiseParallel, ColwiseParallel
 
 
-def apply_pipeline_parallelism(model, world_mesh, parallel_dims, job_config: JobConfig):
+class DummyTransformerLayer(torch.nn.Module):
+    def forward(self, input, freqs_cis):
+        return input
+
+
+class TransformerChunk(torch.nn.Module):
+    def __init__(
+        self,
+        orig_model,  # : Transformer,
+        this_stage_layer_names: List[str],
+        device,
+        input_seqlen: int,
+    ):
+        super().__init__()
+        self.tok_embeddings = None
+
+        # inferring seqlen from forward(input) only works on stage0, bc on later stages
+        # the hidden state input may have reduced seqlen due to TP.  We need to use the
+        # original (full) seqlen for freqs_cis to be correct.
+        self.input_seqlen = input_seqlen
```

Comment:
@wanchaol actually is this a serious bug? i think seq_len might vary from batch to batch, but i was about to treat it as a static quantity.

A problem is that in the PP case, no tensor that is visible from the top level module has the right seq_len - the input tensor has the shape [bsz//#chunks, seqlen // TP], and I dont really want my code to know about TP and multiply that back in.  Its not a problem in the non-PP case, since you have the full model and you can easily store the input seqlen from the top of the forward. 

The solution space seems to be either 
1) make seq_len an output/input of the pipeline stages (maybe have to wrap/unwrap in tensor to make PP happy), so later stages can access the real value

2) slice the self.freqs_cis buffer deeper inside the layer code, after we gather the sharded seq parallel activations and restore the input seqlen

I think 2 is nicer if we can do it easily. but i didnt try it yet to see if its easy